google-http-client-android/src/main/java/com/google/api/client/extensions/android/util/store/FileDataStoreFactory.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.client.extensions.android.util.store;
+
+import com.google.api.client.util.IOUtils;
+import com.google.api.client.util.Maps;
+import com.google.api.client.util.store.AbstractDataStoreFactory;
+import com.google.api.client.util.store.AbstractMemoryDataStore;
+import com.google.api.client.util.store.DataStore;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.logging.Logger;
+
+/**
+ * Thread-safe file implementation of a credential store.
+ *
+ * <p>For security purposes, the file's permissions are set to be accessible only by the file's
+ * owner. Note that Java 1.5 does not support manipulating file permissions, and must be done
+ * manually or using the JNI.
+ *
+ * <p>Note: This class was branched from the primary implementation in google-http-client to allow
+ * the mainline implementation to support Windows file permissions.
+ *
+ * @since 1.36
+ * @author Yaniv Inbar
+ */
+public class FileDataStoreFactory extends AbstractDataStoreFactory {
+
+  private static final Logger LOGGER = Logger.getLogger(FileDataStoreFactory.class.getName());
+
+  /** Directory to store data. */
+  private final File dataDirectory;
+
+  /** @param dataDirectory data directory */
+  public FileDataStoreFactory(File dataDirectory) throws IOException {
+    dataDirectory = dataDirectory.getCanonicalFile();
+    this.dataDirectory = dataDirectory;
+    // error if it is a symbolic link
+    if (IOUtils.isSymbolicLink(dataDirectory)) {
```

Comment:
the object has been modified before the exception is thrown

Reply:
Done. Moved the member assignment after the validations here and in the mainline implementation.